Add -Wextra -Werror to agent build

adalton opened the pull request:
Currently our builds include the -Wall compiler flag, which enables
some, but not all, warnings. This change adds the -Wextra flag, which
enables some additional warnings.

This change also adds -Werror, which will cause the compiler to report
any warning as an error (so now anything that generates a warning will
break the build). This change explicitly omits some pervasive warnings
(e.g., unused parameters).

Note that the change to `syscall_table.c` removed some duplicate initializations.

gianlucaborello commented:
I appreciate the effort and I think this is good. However, in my opinion it is too conservative to have it enabled by default. Sysdig is a fairly popular open source project that is built by several external distribution maintainers with many different compilers, and we don't want to cause too many disruptions to those poor souls.

I would put the `-Werror` under an optional Cmake `option()` (or whatever other mechanism you deem appropriate), disabled by default. For our internal usage, it can be turned on by default via our bootstrap scripts, if we need it.

For example, on my standard box where I do all sysdig development (GCC 8), these are all the warnings I'm getting after this that cause my build to fail: 

```
/home/gianluca/agent/sysdig/userspace/libscap/scap.c: In function ‘scap_open_live_int’:
/home/gianluca/agent/sysdig/userspace/libscap/scap.c:316:22: warning: ‘/dev/sysdig’ directive writing 11 bytes into a region of size between 0 and 1024 [-Wformat-overflow=]
    sprintf(filename, "%s/dev/" PROBE_DEVICE_NAME "%d", scap_get_host_root(), all_scanned_devs);
/home/gianluca/agent/sysdig/userspace/libscap/scap.c:316:25: note: format string is defined here
    sprintf(filename, "%s/dev/" PROBE_DEVICE_NAME "%d", scap_get_host_root(), all_scanned_devs);
/home/gianluca/agent/sysdig/userspace/libscap/scap.c:316:22: note: using the range [-2147483648, 2147483647] for directive argument
    sprintf(filename, "%s/dev/" PROBE_DEVICE_NAME "%d", scap_get_host_root(), all_scanned_devs);
/home/gianluca/agent/sysdig/userspace/libscap/scap.c:316:4: note: ‘sprintf’ output between 13 and 1047 bytes into a destination of size 1024
    sprintf(filename, "%s/dev/" PROBE_DEVICE_NAME "%d", scap_get_host_root(), all_scanned_devs);
/home/gianluca/agent/sysdig/userspace/libscap/scap_fds.c: In function ‘scap_fd_read_unix_sockets_from_proc_fs’:
/home/gianluca/agent/sysdig/userspace/libscap/scap_fds.c:1077:4: warning: ‘strncpy’ specified bound 1024 equals destination size [-Wstringop-truncation]
    strncpy(fdinfo->info.unix_socket_info.fname, token, SCAP_MAX_PATH_SIZE);
/home/gianluca/agent/sysdig/userspace/libscap/scap_userlist.c: In function ‘scap_create_userlist’:
/home/gianluca/agent/sysdig/userspace/libscap/scap_userlist.c:105:4: warning: ‘strncpy’ specified bound 256 equals destination size [-Wstringop-truncation]
    strncpy(handle->m_userlist->users[usercnt].name, p->pw_name, sizeof(handle->m_userlist->users[usercnt].name));
/home/gianluca/agent/sysdig/userspace/libscap/scap_userlist.c:114:4: warning: ‘strncpy’ specified bound 1024 equals destination size [-Wstringop-truncation]
    strncpy(handle->m_userlist->users[usercnt].homedir, p->pw_dir, sizeof(handle->m_userlist->users[usercnt].homedir));
In file included from /home/gianluca/agent/sysdig/userspace/libsinsp/chisel.cpp:31:
In function ‘int tinydir_readfile(const tinydir_dir*, tinydir_file*)’,
    inlined from ‘static void sinsp_chisel::get_chisel_list(std::vector<chisel_desc>*)’ at /home/gianluca/agent/sysdig/userspace/libsinsp/chisel.cpp:1114:20:
/home/gianluca/agent/sysdig/userspace/libsinsp/./third-party/tinydir.h:357:8: warning: ‘char* strcat(char*, const char*)’ accessing 4097 or more bytes at offsets 0 and 4096 may overlap 1 byte at offset 4096 [-Wrestrict]
  strcat(file->path, file->name);
/home/gianluca/agent/sysdig/userspace/libsinsp/ctext.cpp: In function ‘void search_copy(ctext_search*, ctext_search*)’:
/home/gianluca/agent/sysdig/userspace/libsinsp/ctext.cpp:43:56: warning: ‘void* memcpy(void*, const void*, size_t)’ writing to an object of type ‘ctext_search’ {aka ‘struct ctext_search_struct’} with no trivial copy-assignment; use copy-assignment or copy-initialization instead [-Wclass-memaccess]
  memcpy(dst, src, sizeof(ctext_search) - sizeof(string));
In file included from /home/gianluca/agent/sysdig/userspace/libsinsp/ctext.cpp:22:
/home/gianluca/agent/sysdig/userspace/libsinsp/ctext.h:148:16: note: ‘ctext_search’ {aka ‘struct ctext_search_struct’} declared here
 typedef struct ctext_search_struct
/home/gianluca/agent/sysdig/userspace/libsinsp/cursesui.cpp: In member function ‘sysdig_table_action sinsp_cursesui::handle_textbox_input(int)’:
/home/gianluca/agent/sysdig/userspace/libsinsp/cursesui.cpp:2286:28: warning: catching polymorphic type ‘struct sinsp_exception’ by value [-Wcatch-value=]
      catch(sinsp_exception e)
In file included from /home/gianluca/agent/sysdig/userspace/libsinsp/socket_collector.h:27,
                 from /home/gianluca/agent/sysdig/userspace/libsinsp/docker.h:28,
                 from /home/gianluca/agent/sysdig/userspace/libsinsp/docker.cpp:27:
In function ‘static int socket_data_handler<T>::ssl_key_password_cb(char*, int, int, void*) [with T = docker]’,
    inlined from ‘static int socket_data_handler<T>::ssl_key_password_cb(char*, int, int, void*) [with T = docker]’ at /home/gianluca/agent/sysdig/userspace/libsinsp/socket_handler.h:929:13:
/home/gianluca/agent/sysdig/userspace/libsinsp/socket_handler.h:936:11: warning: ‘char* strncpy(char*, const char*, size_t)’ output truncated before terminating nul copying as many bytes from a string as its length [-Wstringop-truncation]
    strncpy(buf, (const char*)pass, pass_len);
/home/gianluca/agent/sysdig/userspace/libsinsp/socket_handler.h: In static member function ‘static int socket_data_handler<T>::ssl_key_password_cb(char*, int, int, void*) [with T = docker]’:
/home/gianluca/agent/sysdig/userspace/libsinsp/socket_handler.h:934:42: note: length computed here
    int pass_len = static_cast<int>(strlen((char*)pass));
In file included from /home/gianluca/agent/sysdig/userspace/libsinsp/socket_collector.h:27,
                 from /home/gianluca/agent/sysdig/userspace/libsinsp/k8s_handler.h:27,
                 from /home/gianluca/agent/sysdig/userspace/libsinsp/k8s_handler.cpp:24:
In function ‘static int socket_data_handler<T>::ssl_key_password_cb(char*, int, int, void*) [with T = k8s_handler]’,
    inlined from ‘static int socket_data_handler<T>::ssl_key_password_cb(char*, int, int, void*) [with T = k8s_handler]’ at /home/gianluca/agent/sysdig/userspace/libsinsp/socket_handler.h:929:13:
/home/gianluca/agent/sysdig/userspace/libsinsp/socket_handler.h:936:11: warning: ‘char* strncpy(char*, const char*, size_t)’ output truncated before terminating nul copying as many bytes from a string as its length [-Wstringop-truncation]
    strncpy(buf, (const char*)pass, pass_len);
/home/gianluca/agent/sysdig/userspace/libsinsp/socket_handler.h: In static member function ‘static int socket_data_handler<T>::ssl_key_password_cb(char*, int, int, void*) [with T = k8s_handler]’:
/home/gianluca/agent/sysdig/userspace/libsinsp/socket_handler.h:934:42: note: length computed here
    int pass_len = static_cast<int>(strlen((char*)pass));
                                    ~~~~~~^~~~~~~~~~~~~
/home/gianluca/agent/sysdig/userspace/libsinsp/threadinfo.cpp: In member function ‘void sinsp_threadinfo::fd_to_scap(scap_fdinfo*, sinsp_fdinfo_t*)’:
/home/gianluca/agent/sysdig/userspace/libsinsp/threadinfo.cpp:1153:10: warning: ‘char* strncpy(char*, const char*, size_t)’ specified bound 1024 equals destination size [-Wstringop-truncation]
   strncpy(dst->info.regularinfo.fname, src->m_name.c_str(), SCAP_MAX_PATH_SIZE);
/home/gianluca/agent/sysdig/userspace/libsinsp/threadinfo.cpp:1149:10: warning: ‘char* strncpy(char*, const char*, size_t)’ specified bound 1024 equals destination size [-Wstringop-truncation]
   strncpy(dst->info.unix_socket_info.fname, src->m_name.c_str(), SCAP_MAX_PATH_SIZE);
/home/gianluca/agent/sysdig/userspace/libsinsp/threadinfo.cpp:1165:10: warning: ‘char* strncpy(char*, const char*, size_t)’ specified bound 1024 equals destination size [-Wstringop-truncation]
   strncpy(dst->info.fname, src->m_name.c_str(), SCAP_MAX_PATH_SIZE);
In function ‘void sinsp_threadinfo::fd_to_scap(scap_fdinfo*, sinsp_fdinfo_t*)’,
    inlined from ‘sinsp_thread_manager::dump_threads_to_file(scap_dumper_t*)::<lambda(sinsp_threadinfo&)>’ at /home/gianluca/agent/sysdig/userspace/libsinsp/threadinfo.cpp:1611:21,
    inlined from ‘static _Res std::_Function_handler<_Res(_ArgTypes ...), _Functor>::_M_invoke(const std::_Any_data&, _ArgTypes&& ...) [with _Res = bool; _Functor = sinsp_thread_manager::dump_threads_to_file(scap_dumper_t*)::<lambda(sinsp_threadinfo&)>; _ArgTypes = {sinsp_threadinfo&}]’ at /usr/include/c++/8/bits/std_function.h:282:44:
/home/gianluca/agent/sysdig/userspace/libsinsp/threadinfo.cpp:1153:10: warning: ‘char* strncpy(char*, const char*, size_t)’ specified bound 1024 equals destination size [-Wstringop-truncation]
   strncpy(dst->info.regularinfo.fname, src->m_name.c_str(), SCAP_MAX_PATH_SIZE);
/home/gianluca/agent/sysdig/userspace/libsinsp/threadinfo.cpp:1149:10: warning: ‘char* strncpy(char*, const char*, size_t)’ specified bound 1024 equals destination size [-Wstringop-truncation]
   strncpy(dst->info.unix_socket_info.fname, src->m_name.c_str(), SCAP_MAX_PATH_SIZE);
/home/gianluca/agent/sysdig/userspace/libsinsp/threadinfo.cpp:1165:10: warning: ‘char* strncpy(char*, const char*, size_t)’ specified bound 1024 equals destination size [-Wstringop-truncation]
   strncpy(dst->info.fname, src->m_name.c_str(), SCAP_MAX_PATH_SIZE);
/home/gianluca/agent/sysdig/userspace/sysdig/csysdig.cpp: In function ‘sysdig_init_res csysdig_init(int, char**)’:
/home/gianluca/agent/sysdig/userspace/sysdig/csysdig.cpp:791:27: warning: catching polymorphic type ‘struct sinsp_exception’ by value [-Wcatch-value=]
     catch(sinsp_exception e)
/home/gianluca/agent/sysdig/userspace/sysdig/sysdig.cpp: In function ‘sysdig_init_res sysdig_init(int, char**)’:
/home/gianluca/agent/sysdig/userspace/sysdig/sysdig.cpp:1431:27: warning: catching polymorphic type ‘struct sinsp_exception’ by value [-Wcatch-value=]
     catch(sinsp_exception e)
```

gianlucaborello commented on CMakeLists.txt:
```diff
@@ -58,8 +58,11 @@ if(NOT WIN32)
 
-	set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -ggdb")
-	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -ggdb --std=c++0x")
+    set(CMAKE_SUPPRESSED_WARNINGS "-Wno-unused-parameter -Wno-missing-field-initializers -Wno-sign-compare -Wno-type-limits -Wno-implicit-fallthrough -Wno-format-truncation")
```
I believe this is indented using spaces, whereas the whole file is using tabs.

adalton commented:
Hi @gianlucaborello .  Thank you for the feedback, that makes sense.  I've incorporated your suggestion to make building with `-Werror` optional (off by default) for sysdig.  In my agent review (https://github.com/draios/agent/pull/1098), I've updated the ` bootstrap-sysdig` script to enable that option.  Does that seem reasonable to you, or are there use cases for `bootstrap-sysdig` where having that enabled doesn't make sense?  Thanks!

gianlucaborello commented:
Much better.

If it was for me, I would like the ability to override the default behavior in both `bootstrap-sysdig` and `bootstrap-agent` via environment variables (kind of what we do for `WORKDIR` and `BUILD_BPF`, which I use extensively). My rationale is purely selfish, since most of the sysdig/agent work I do doesn't happen on a CentOS 6 container, it's just not practical for me. Since I can't give a solid explanation besides my personal workflow, feel free to ignore it if it's too much work, I can always just change it in my working directory.

By the way, there's a space in `if (BUILD_WARNINGS_AS_ERRORS)`, I don't know if our whitespace coding convention also applies to CMakeLists.txt files, but judging from the overwhelming amount of `if(` vs `if (` in all our CMakeLists.txt across the repos, I'd say it does. Feel free to change it if you think it makes sense.

Also, since this change extends to OSS sysdig, you should probably look if it makes sense to apply it to OSS falco as well, you'll notice the various build scripts and cmake files are loosely kept in sync among the 3 projects.

adalton commented:
HI @gianlucaborello .  I implemented the change that you suggested to enable users of the bootstrap scripts (both for sysdig and falco) to override `BUILD_WARNINGS_AS_ERRORS` via the environment.  I've updated the agent PR and opened a oss-falco PR at https://github.com/falcosecurity/falco/pull/477 (I can't seem to explicitly add reviewers to that PR, maybe I'm not on the "blessed" list for that repo?)

Making the change to falco required a small change where we were trying to throw exceptions from a destructor; I changed that to a log+assert (throwing the exception would have resulted in a call to `terminate()`).

gianlucaborello commented:
For falco, I know the process changed recently due to CNCF and we probably don't all have the original privileges in the new organization (I don't either), so I suggest you discuss this in the falco slack channels.
